### DISCUSSION TITLE
How can I specify setuptools version?

### DISCUSSION THREAD

**Discussion by javenwang** [pdm-project/pdm#1219] [Q&A]
I have a dependency which depends on [DEAP](https://github.com/DEAP/deap/). DEAP requires setuptools<58.0. So I wonder is there any way to specify the setuptools version when running `pdm add deap`?

**Comment by frostming** [comment]
Unless it specifies the requirement version range in `pyproject.toml` per [PEP 518](https://peps.python.org/pep-0518/#specification)'s specification. Maybe it is time to make a PR for them?

As a workaround, you can specify it in *your* project dependencies and use `pdm add --no-isolation deap`.

```toml
[project]
...
dependencies = ["setuptools<58"]
```

**Reply by javenwang** [reply]
thanks for your reply. do i have to put it in `dependencies`? i mean `setuptools` is required only in install/build phase.

i just tried your way but still not working:
```
$ pdm add --no-isolation -v deap
...
Preparing isolated env for PEP 517 build...
...
Collecting setuptools>=40.8.0
...
Installing collected packages: wheel, setuptools
Successfully installed setuptools-63.1.0 wheel-0.37.1
error in deap setup command: use_2to3 is invalid.
Install deap 1.3.1 failed
```

PDM still installed setuptools-63.1.0 even though i have setuptools-57.5.0 installed.

**Reply by frostming** [reply]
Hmm, the workaround only works for extra requirements other than setuptools itself.

**Comment by hcoona** [comment]
watching
